src/GitHub.Api/Git/RepositoryManager.cs

```diff
@@ -331,13 +331,21 @@ public ITask DiscardChanges(GitStatusEntry[] gitStatusEntries)
                     if (itemsToRevert.Any())
                     {
                         gitDiscardTask = GitClient.Discard(itemsToRevert);
-                        task.Then(gitDiscardTask);
+                        task
+                            .Then(gitDiscardTask)
+                            // we're appending a new continuation, we need to reset the finally handler
+                            // so it runs after the discard task
```

Comment:
By adding to the then we are bypassing the finally that is going to get attached to this task?

Reply:
https://github.com/github-for-unity/Unity/blob/8b07cb2e7046bf15dde863b59b94c4aeec361ead/src/GitHub.Api/Git/RepositoryManager.cs#L338-L340

You tweaked the message here, but what happened to the original `Finally` that `HookupHandlers` attached to.

Reply:
The Finally has already been appended to the task. This code executes in the body of the task, where HookupHandlers executes before the task runs. The Finally won't be added to the new task that we're appending to the currently executing task, because Finally and Catch handlers only propagate up the chain, not down the chain.

If chain Y+Finally is appended to A->B, the finally handler gets set on all the tasks up to the top so that if any of them fail, it will always get called. This means the chain will be A+Finally->B+Finally->Y+Finally. This is mainly because it's simpler (I only have to deal with one finally handler per task, not many), and 99% of the times we finish building a chain by appending the one that has the finally handler (or we set up the finally handler after we're done building the chain).

In this particular case, though, we're hacking into the chain mid-execution to insert something in between. The chain is A+Finally and we're going A+Finally->B. That means that as soon as A is done, it looks and sees that B is there and calls B. When B is done, it doesn't see any finally handler on B so it doesn't call it. Which is why I'm adding the Finally here, to make sure it gets set at the end.

This isn't really a fix, it's a hack - there are proper ways to fix this. Most likely I would change the way I store finally handlers to always be at the root of the chain, and just go up the chain to call them whenever the chain is finished. That might cause some subtle differences in behaviour that need some more thorough testing.

Reply:
Okay, I get it.